logging/logging_test.go

```diff
@@ -1358,9 +1355,90 @@ func TestRedirectOutputFormats(t *testing.T) {
 	}
 }
 
-func ExampleRedirectAsJson_withStdout() {
-	// use previously created client
+func TestInstrumentationIngestion(t *testing.T) {
+	var got []*logpb.LogEntry
+
+	tests := []struct {
+		name          string
+		ingestionFlag bool
+		want          int
+	}{
+		{
+			name:          "first time log",
+			ingestionFlag: true,
+			want:          2,
+		},
+		{
+			name:          "any non first log",
+			ingestionFlag: false,
+			want:          1,
+		},
+	}
+	// setup fake client
+	client, err := fakeClient("projects/test", func(e *logpb.WriteLogEntriesRequest) {
+		got = e.GetEntries()
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+	iiStatus := internal.IngestInstrumentation()
+
+	entry := &logging.Entry{Severity: logging.Info, Payload: "test string"}
+	logger := client.Logger("test-instrumentation")
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			got = nil
+			internal.SetIngestInstrumentation(tc.ingestionFlag)
+			err := logger.LogSync(context.TODO(), *entry)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if len(got) != tc.want {
+				t.Errorf("got(%v),want(%v):\n", got, tc.want)
+			}
+			diagnosticEntry := false
+			for _, ent := range got {
+				if internal.LogIDFromPath("projects/test", ent.LogName) == "diagnostic-log" {
+					diagnosticEntry = true
+					break
+				}
+			}
+			if tc.ingestionFlag != diagnosticEntry {
+				t.Errorf("instrumentation entry misplaced: got(%v),want(%v):\n", diagnosticEntry, tc.ingestionFlag)
+			}
+		})
+	}
+	internal.SetIngestInstrumentation(iiStatus)
+}
+
+func TestInstrumentationWithRedirect(t *testing.T) {
+	want := []string{
+		// do not format the string to preserve expected new-line between messages
+		`{"message":"test string","severity":"INFO","timestamp":"seconds:1000"}
+{"message":{"logging.googleapis.com/diagnostic":{"instrumentation_source":[{"name":"go","version":"` + internal.Version + `"}],"runtime":"` + internal.VersionGo() + `"}},"severity":"DEFAULT","timestamp":"seconds:1000"}`,
+		`{"message":"test string","severity":"INFO","timestamp":"seconds:1000"}`,
+	}
+	entry := &logging.Entry{Severity: logging.Info, Payload: "test string"}
+	buffer := &strings.Builder{}
+	logger := client.Logger("test-redirect-output", logging.RedirectAsJSON(buffer))
+	iiStatus := internal.IngestInstrumentation()
+	internal.SetIngestInstrumentation(true)
+	for i, _ := range want {
```

Comment:
nit: can remove second return value completly from range instead of using underscore.